java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java

```diff
@@ -476,11 +507,12 @@ private void allocateBytes(final long valueBufferSize, final int valueCount) {
   }
 
   /* allocate offset buffer */
-  private void allocateOffsetBuffer(final long size) {
+  private ArrowBuf allocateOffsetBuffer(final long size) {
```

Comment:
Is the intent of this change to avoid changing the existing offset buffer, and only change it for the exported version? Then I must ask

(1) why? If we're allocating the buffer, we may as well use it for the vector as well.
(2) is this safe in terms of leaking memory? since apparently nothing will hold on to the buffer after allocating it above

Reply:
For 2, this is only for exporting the vectors through C Data Interface. By it means, it is intended to be released by the importer side. At Java Arrow, I remember ArrayExporter only increases the reference count for the buffers to be exported to prevent them to be released automatically.

Reply:
For 1, they are allocated for using only at imported side. I think as per discussed with @pitrou above, the idea is these empty buffers exist internally and it is okay.

Reply:
I think Antoine's idea was that you could statically allocate one such buffer to use in such situations, if you're talking about this

> As long as data stays internal to Arrow Java, the missing buffer is ok, but for purposes of exchanging data, a proper offsets buffers should be exported (this one can easily be statically allocated, by the way).

Reply:
And yes, it seems ArrayExporter will increase the reference count by 1. But since it starts with a reference count of 1, that means these buffers will never be freed.

Reply:
> And yes, it seems ArrayExporter will increase the reference count by 1. But since it starts with a reference count of 1, that means these buffers will never be freed.

No. There is callback in exported structured that is used to release buffer. The responsibility of calling the callback is on imported side (for example, Rust arrow, for our case). Once the imported side doesn't need the vector anymore, it will trigger the callback to release the buffer allocated in exported side.

Reply:
Which will decrement the refcount by 1, but the refcount of this buffer will be 2, right? It starts at 1 when you allocate it, and is increased by 1 when you export it. Then Rust calls the callback and the refcount is decremented by 1. It will not get freed.

Reply:
See:

https://github.com/apache/arrow/blob/0dbbd43ca9133912d1809394727784560cc5e797/java/c/src/main/cpp/jni_wrapper.cc#L142
https://github.com/apache/arrow/blob/0dbbd43ca9133912d1809394727784560cc5e797/java/c/src/main/java/org/apache/arrow/c/ArrayExporter.java#L57

Reply:
I'm well aware of how C Data works. I'm asking about how the reference counting is supposed to work here.

Reply:
Closing the vector here won't affect the reference count of this buffer because it explicitly isn't set as the vector's offset buffer.

Reply:
Okay, I got your point. I don't familiar with Java Arrow buffer allocation mechanism. I traced how it is allocated now so the ref count is increased automatically once it is allocated.

Reply:
> I think Antoine's idea was that you could statically allocate one such buffer to use in such situations, if you're talking about this
> 
> > As long as data stays internal to Arrow Java, the missing buffer is ok, but for purposes of exchanging data, a proper offsets buffers should be exported (this one can easily be statically allocated, by the way).

Hmm, I think what Antoine means is, we should export a proper offset buffer (instead of empty one currently) for exporting purpose, isn't?

Reply:
Yes, we can export a proper offset buffer, **but all such cases can share the same buffer** because the data is always the same

Reply:
I'm saying that once you export the buffer, you might as well set it to be the offset buffer? Not when you create the vector. (That may also be surprising to users, sure.) But it would be the only way to avoid this memory leak.

Reply:
Hmm, I think I got your point. But I am wondering if it makes sense. I doubt that after exporting an array, it still makes sense to hold and modify it at exported side.

Especially,

https://arrow.apache.org/docs/format/CDataInterface.html#mutability

> Both the producer and the consumer SHOULD consider the exported data (that is, the data reachable through the buffers member of ArrowArray) to be immutable, as either party could otherwise see inconsistent data while the other is mutating it.

Reply:
If you insist that the offset buffer must keep same buffer between Java Arrow and exported side. The only way is to set it (the one value offset) to the offset of the vector. But as I said before, it will cause a lot inconsistency in Java Arrow.

I don't think an inconsistent offset (sometimes it is empty sometimes it is one value buffer) is a good idea. Very confused and easy to cause bugs.

Reply:
> Now the created one value buffer only has one ref count for imported side.

I've gone over this: when a buffer is allocated, it starts with a refcount of 1. As you yourself pointed out, exporting it increases the refcount. And when the consumer calls the release callback, the refcount is only decreased by 1. So the refcount does not reach 0.

Reply:
This works for the other buffers of the vector because the vector holds the buffer, so when the vector itself is closed it decrements the refcounts of the buffers, so they reach 0. But for the buffer which we allocate on the fly here, that is not true.

Reply:
> I don't think an inconsistent offset (sometimes it is empty sometimes it is one value buffer) is a good idea. Very confused and easy to cause bugs.

Code shouldn't make assumptions because we've already documented that those assumptions are wrong. It would be equally true for code working with imported arrays: they may have a one value buffer.

Reply:
The mutability ship has sailed for the Java implementation. It would be no different for any array here no matter what we do.

Reply:
> > Now the created one value buffer only has one ref count for imported side.
> 
> I've gone over this: when a buffer is allocated, it starts with a refcount of 1. As you yourself pointed out, exporting it increases the refcount. And when the consumer calls the release callback, the refcount is only decreased by 1. So the refcount does not reach 0.

Hmm, don't you see my latest commit? It doesn't increase ref count when exporting.

Reply:
This is making the problem worse for all non-offset buffers (which now get an extra refcount) while not changing it for the offset buffer (because, again, please see the sequence of events that I keep outlining above...)

Reply:
> > I don't think an inconsistent offset (sometimes it is empty sometimes it is one value buffer) is a good idea. Very confused and easy to cause bugs.
> 
> Code shouldn't make assumptions because we've already documented that those assumptions are wrong. It would be equally true for code working with imported arrays: they may have a one value buffer.

Hmm, but this is the current code status...

You can see how these var-size vectors handle offset buffer, they assume that the initial capacity of offset buffer is 0. That's why I got some many test failure after I changed the initial offset buffer.

Reply:
To move this forward instead of arguing continuously. I would like confirm that if the following approach is acceptable to you.

1. Not change current initial offset buffer when creating vector, i.e., they are empty as usual.
2. Allocate one value offset buffer when exporting vector and replace the existing (empty) offset buffer.
3. For the places in var-size vectors which have the assumption about initial offset buffer capacity (zero), changing it or throwing some error (as these places are going to re-allocate offset buffer).

WDYT?

Reply:
Or, alternatively you would like to work on this if you have time and are willing to fix it soon (as we are blocked by this somehow). Although we could apply currently fix internally to unblock us, I would like to have consistent code between fork and open source and avoid resolving conflict later.

WDYT?

Reply:
>> Hmm, don't you see my latest commit? It doesn't increase ref count when exporting.
>> Or I miss anything?
> This is making the problem worse for all non-offset buffers (which now get an extra refcount) while not changing it for the offset buffer (because, again, please see the sequence of events that I keep outlining above...)

I know what you meant, and I guess you don't see committed code but only the comment, so there is confusion.

In latest commit, I don't increase ref count when exporting offset buffer. For non-offset buffers, of course their refcounts are increased as before.

Reply:
@viirya would it be possible to do a simple end-to-end experiment and log how the refcout varies in the original source vs the new change (at least the new change) and see what is actually happening. So if we have some call cancelling a mistake we made in the refcount could be thus reflected and clear our doubts. Is this a viable?

Reply:
Hmm, this is a good suggestion. I'm not sure if it is viable now. I will try to find time looking into it.

Reply:
I understand. Let me also evaluate this a bit thoroughly.

Reply:
@vibhatha I've added some more checks (buffer ref count, allocated memory) into the existing roundtrip tests of C data interface. Please take a look. Thanks.

Reply:
I will take a look at this as soon as possible.

Reply:
Hi @vibhatha, sorry for pinging you. Do you have a chance to look at this recently? One important feature is blocked by the issue and we are looking forward to move this into Java Arrow library. Thank you. 🙏

Reply:
@viirya I couldn't take a look. But let me see if I can allocate some time to do this probably earlier next week?

Reply:
@vibhatha It will be great! Thank you so much!